contrib/verify-commits/trusted-keys

```diff
@@ -1,6 +1,7 @@
-71A3B16735405025D447E8F274810B012346C9A6
-B8B3F1C0E58C15DB6A81D30C3648A882F4316B9B
-E777299FC265DD04793070EB944D35F9AC3DB76A
-D1DBF2C4B96F2DEBF4C16654410108112E7EA81F
-152812300785C96444D3334D17565732E08E5E41
-6B002C6EA3F91B1B0DF0C9BC8F617F1200A6D25C
+key,since,until
```

Comment:
Not sure if it makes sense to document the last merge commit here.

Keeping the key will also mean that the script will break again once the key naturally expires or is revoked. So I guess you'd also have to document a faketime where the script still passes? Moreover, the key might be purged from keyservers if it is expired/revoked, so you may also have to document the full key here. Not sure if we want this.

It seems easier to just bump the trusted root, like it was done before, see d4b3dc5b0a726cc4cc7a8467be43126e78f841cf

The trusted root hash commits to all previous commits, so there shouldn't be any downside, compared to listing a subset of commits prior to the trusted root hash.

Of course anyone is free to archive previous keys as long as they want and re-run gpg on previous commits, with or without faketime, as often as they want. But doing that once as part of the review process of bumping the trusted root should be enough. I don't see a need to maintain every key forever.

Overall I think, the script should ideally be easy to understand and use, so that many users can read and use it. It is already hard enough to properly run the script (just one example: https://github.com/bitcoin/bitcoin/pull/25197#issuecomment-1137097923 ; happy to provide more examples), so adding even more complexity and failure modes will mean that likely no one is running it in practice, let alone understand it and be able to act on errors if there are any.

Reply:
For me, the question was how comfortable are we with a really recent trusted root? Suppose an active maintainer needs to be removed on very short notice, e.g. keys and account is compromised. We could drop the key and update the trusted root to the current `HEAD`, but are we okay with doing that? With this change, we could just set the `until` commit rather than removing the key.

My thinking was that the trusted root shouldn't be updated ever, and really we should make it possible to move it even earlier so that older commits signed by previous maintainers can still be verified. But maybe that's not how people think of verify-commits.

Reply:
> But maybe that's not how people think of verify-commits.

Yeah opinions may differ on this. Though, I think there is no easy way out of manually reviewing (and manually taking over! [1]) changes to this script and the data it uses. So optimizing for an easy and repeatable review process should be priority, to allow as many people to repeat it as possible. Otherwise, if the review process is different for each occasion, it is impossible to document and hard to follow. I do think it probably makes sense to allow for one-off rare exceptions (unclean merge, unsigned merge, wrong sha512, ...), which can be used during the review process of bumping the trusted root, but I think will complicate stuff if they are needed to be maintained for the whole history at the HEAD commit forever. (Note that anyone can check out an earlier version to get them for a previous commit, even if they are removed from HEAD)

> We could drop the key and update the trusted root to the current HEAD, but are we okay with doing that?

I'd say yes. There is no way out of doing a change on short notice. And then, whether reviewers review the `until` commit, or the trusted root, which will be equal to the `until` commit, shouldn't make a difference?




[1] Blindly checking out the script on the fetched/downloaded/merged git commit is next to useless, unless you are the CI doing a smoke test.

Reply:
I've dropped this commit for now.